Automattic/wp-calypso
# Store: Update `ProductSearch` component for better variation support, multiple selections

See #15681. We need a reusable product search component for orders and promotions, so I've pulled together the code from what's currently in the promotions section and added variation support.

We now have a component that renders out all the products in a store, which can be filtered by searching. Products with variations are listed with a Count component (mostly to take up the space where the checkbox should be), and a Customize button, which opens the variations form.

<img width="600" alt="list" src="https://user-images.githubusercontent.com/541093/32455496-77bd6532-c2f0-11e7-8d69-cc8fe1b4d420.png">

As products are selected in the form, they become selected rows in the list. Unchecking them deletes them from the list. They're also displayed with the correct variation info - price and image.

<img width="602" alt="list-selected" src="https://user-images.githubusercontent.com/541093/32455495-77ad6bdc-c2f0-11e7-8e92-a1d645d2106a.png">

This component also supports enforcing a single selection. This uses radio buttons, instead of checkboxes.

<img width="595" alt="single" src="https://user-images.githubusercontent.com/541093/32455494-779cb8d2-c2f0-11e7-8452-a6c073fc600e.png">

It only allows for one variation to be selected at a time, and will replace the currently-selected variation the new one if it's changed.

<img width="591" alt="single-selected" src="https://user-images.githubusercontent.com/541093/32455493-778fd2fc-c2f0-11e7-961f-395b9492a11c.png">

You can also [check out the new README for examples/expected props](https://github.com/Automattic/wp-calypso/blob/add/store-new-product-search/client/extensions/woocommerce/components/product-search/README.md).

ℹ️ Note: This PR also updates the `per_page` for the variations API request. It was set to 10, so my Mug example with 12 variations (3 sizes, 4 colors) was not fully loading here or on the product edit screen.

⚠️ Known issue: currently when searching, selected simple products stay in the list even if they don't match (a holdover from the appliesTo component). This isn't true for variations right now, but I wanted to finally get this PR up before it took even longer.

❓ Question: Currently the text says that the user can add the "base product" without picking variations, but I haven't added that functionality yet. Is that something we should support?

**To test**

This isn't in use anywhere real yet, but you can still test it by checking out the order create screen

- View order create page by visiting `/store/order/:site`
- You can find the search under "Add products to the order"
- Make sure all your products load in (they come in either 50 or 100 at a time, so if you have a lot of products they might not all be there immediately)
- Try selecting some products
- Click around the variations, select and unselect them
- You can edit `order-create/index.js` and add the `singular` prop to see the single-select view
- Additional testing of the Products edit screen to make sure all variations load in the variations table would also help 👍 

As usual, if anyone has better copy I'm happy to update anything here 🙂
